File: Examples/ShapeAreas/ShapeAreas.cs
// SPDX-FileCopyrightText: 2026 Progmasoft <support@progmasoft.com>
// SPDX-License-Identifier: MPL-2.0 WITH AdditionRef-Progmasoft-Exception-1.0

using System;

namespace Examples.ShapeAreas;

public abstract record Shape;
public sealed record Circle(double Radius) : Shape;
public sealed record Rectangle(double Width, double Height) : Shape;

public static class ShapeAreas
{
    private static double Area(Shape shape) => shape switch
    {
        Circle circle => Math.PI * circle.Radius * circle.Radius,
        Rectangle rectangle => rectangle.Width * rectangle.Height,
        _ => throw new ArgumentOutOfRangeException(nameof(shape)),
    };

    public static void Main()
    {
        Shape[] shapes = [new Circle(3), new Rectangle(4, 5)];
        foreach (var shape in shapes)
        {
            Console.WriteLine($"{Area(shape):F2}");
        }
    }
}

File: Examples/ShapeAreas/ShapeAreas.java
// SPDX-FileCopyrightText: 2026 Progmasoft <support@progmasoft.com>
// SPDX-License-Identifier: MPL-2.0 WITH AdditionRef-Progmasoft-Exception-1.0

public final class ShapeAreas {
    private ShapeAreas() {}

    private sealed interface Shape permits Circle, Rectangle {}
    private record Circle(double radius) implements Shape {}
    private record Rectangle(double width, double height) implements Shape {}

    private static double area(Shape shape) {
        return switch (shape) {
            case Circle circle -> Math.PI * circle.radius() * circle.radius();
            case Rectangle rectangle -> rectangle.width() * rectangle.height();
        };
    }

    public static void main(String[] args) {
        Shape[] shapes = {new Circle(3), new Rectangle(4, 5)};
        for (var shape : shapes) {
            System.out.printf("%.2f%n", area(shape));
        }
    }
}
